crates/criticalup-core/src/state.rs

```diff
@@ -188,20 +189,29 @@ impl State {
         Ref::map(self.inner.borrow(), |v| &v.repr.installations)
     }
 
+    #[tracing::instrument(level = "trace", skip_all)]
     pub fn all_binary_proxy_names(&self) -> Vec<PathBuf> {
-        let mut result: Vec<_> = self
-            .inner
-            .borrow()
-            .repr
-            .installations
-            .values()
-            .flat_map(|installation| installation.binary_proxies.keys())
-            .cloned()
-            .collect();
-
-        result.sort();
-        result.dedup();
-        result
+        let state = self.inner.borrow();
+
+        let mut all_proxies = vec![];
+        for (id, installation) in &state.repr.installations {
+            let span = span!(Level::TRACE, "installation", %id);
+            let _entered = span.enter();
+
+            let installation_proxies = &installation.binary_proxies;
+            if installation_proxies.is_empty() {
+                tracing::trace!("No proxies for installation");
+            } else {
+                for proxy in installation_proxies.keys() {
+                    tracing::trace!(proxy = %proxy.display(), "Found proxy for installation");
+                    all_proxies.push(proxy.clone())
+                }
+            }
```

Comment:
Nice! Not only does it include tracing but also make code more readable and easier to reason.